frankenstein/chunk_store.go

```diff
@@ -344,125 +369,234 @@ func (c *AWSChunkStore) Get(ctx context.Context, from, through model.Time, match
-				return
-			}
-			parts := strings.SplitN(*rangeValue, ":", 2)
-			if len(parts) != 2 {
-				log.Errorf("Invalid item: %v", item)
-				incomingErrors <- err
-				return
-			}
-			chunkValue := item[chunkKey].B
-			if rangeValue == nil {
-				log.Errorf("Invalid item: %v", item)
-				incomingErrors <- err
-				return
-			}
-			chunk := wire.Chunk{
-				ID: parts[1],
-			}
-			if err := json.Unmarshal(chunkValue, &chunk); err != nil {
-				log.Errorf("Invalid item: %v", item)
-				incomingErrors <- err
-				return
-			}
-			chunkSet = append(chunkSet, chunk)
+func (c *AWSChunkStore) lookupChunksForMatcher(userID string, hour int64, metricName model.LabelValue, matcher *metric.LabelMatcher) (wire.ChunksByID, error) {
+	hashValue := hashValue(userID, hour, metricName)
+	var rangeMinValue, rangeMaxValue []byte
+	if matcher.Type == metric.Equal {
+		nextValue := model.LabelValue(next(string(matcher.Value)))
+		rangeMinValue = rangeValue(matcher.Name, matcher.Value, "")
+		rangeMaxValue = rangeValue(matcher.Name, nextValue, "")
+	} else {
+		nextLabel := model.LabelName(next(string(matcher.Name)))
+		rangeMinValue = rangeValue(matcher.Name, "", "")
+		rangeMaxValue = rangeValue(nextLabel, "", "")
+	}
+
+	var resp *dynamodb.QueryOutput
+	err := timeRequest("Query", dynamoRequestDuration, func() error {
+		var err error
+		resp, err = c.dynamodb.Query(&dynamodb.QueryInput{
+			TableName: aws.String(c.tableName),
+			KeyConditions: map[string]*dynamodb.Condition{
+				hashKey: {
+					AttributeValueList: []*dynamodb.AttributeValue{
+						{S: aws.String(hashValue)},
+					},
+					ComparisonOperator: aws.String("EQ"),
+				},
+				rangeKey: {
+					AttributeValueList: []*dynamodb.AttributeValue{
+						{B: rangeMinValue},
+						{B: rangeMaxValue},
+					},
+					ComparisonOperator: aws.String("BETWEEN"),
+				},
+			},
+			ReturnConsumedCapacity: aws.String(dynamodb.ReturnConsumedCapacityTotal),
+		})
+		return err
+	})
+	if resp.ConsumedCapacity != nil {
+		dynamoConsumedCapacity.WithLabelValues("Query").
+			Add(float64(*resp.ConsumedCapacity.CapacityUnits))
+	}
+	if err != nil {
+		log.Errorf("Error querying DynamoDB: %v", err)
+		return nil, err
+	}
+
+	chunkSet := wire.ChunksByID{}
+	for _, item := range resp.Items {
+		rangeValue := item[rangeKey].B
+		if rangeValue == nil {
+			log.Errorf("Invalid item: %v", item)
+			return nil, err
+		}
+		label, value, chunkID, err := parseRangeValue(rangeValue)
+		if err != nil {
+			log.Errorf("Invalid item: %v", item)
+			return nil, err
+		}
+		chunkValue := item[chunkKey].B
+		if rangeValue == nil {
+			log.Errorf("Invalid item: %v", item)
+			return nil, err
 		}
-		chunkSets = append(chunkSets, chunkSet)
+		chunk := wire.Chunk{
+			ID: chunkID,
+		}
+		if err := json.Unmarshal(chunkValue, &chunk); err != nil {
+			log.Errorf("Invalid item: %v", item)
+			return nil, err
+		}
+		if label != matcher.Name || !matcher.Match(value) {
+			log.Debugf("Dropping unexpected", chunk.Metric)
```

Comment:
I should remove the log line really; its expected that we'll hit this quite a lot in normal operation.

Will add a counter.